// @giphy/react-native-sdk/lib/module/dto/giphyRendition.js
export const GiphyRendition = {
  Original: 'original',
  OriginalStill: 'original_still',
  Preview: 'preview',
  Looping: 'looping',
  FixedHeight: 'fixed_height',
  FixedHeightStill: 'fixed_height_still',
  FixedHeightDownsampled: 'fixed_height_downsampled',
  FixedHeightSmall: 'fixed_height_small',
  FixedHeightSmallStill: 'fixed_height_small_still',
  FixedWidth: 'fixed_width',
  FixedWidthStill: 'fixed_width_still',
  FixedWidthDownsampled: 'fixed_width_downsampled',
  FixedWidthSmall: 'fixed_width_small',
  FixedWidthSmallStill: 'fixed_width_small_still',
  Downsized: 'downsized',
  DownsizedSmall: 'downsized_small',
  DownsizedMedium: 'downsized_medium',
  DownsizedLarge: 'downsized_large',
  DownsizedStill: 'downsized_still'
};
export const GiphyClipsRendition = {
  Original: GiphyRendition.Original,
  OriginalStill: GiphyRendition.OriginalStill,
  FixedHeight: GiphyRendition.FixedHeight,
  FixedHeightStill: GiphyRendition.FixedHeightStill,
  FixedHeightDownsampled: GiphyRendition.FixedHeightDownsampled,
  FixedWidth: GiphyRendition.FixedWidth,
  FixedWidthStill: GiphyRendition.FixedWidthStill,
  FixedWidthDownsampled: GiphyRendition.FixedWidthDownsampled,
  DownsizedMedium: GiphyRendition.DownsizedMedium,
  DownsizedLarge: GiphyRendition.DownsizedLarge
};
